src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs

```diff
@@ -0,0 +1,755 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Marshalling;
+using System.Runtime.Versioning;
+using System.Text;
+using EditorBrowsableAttribute = System.ComponentModel.EditorBrowsableAttribute;
+using EditorBrowsableState = System.ComponentModel.EditorBrowsableState;
+
+#pragma warning disable 0809  //warning CS0809: Obsolete member 'ReadOnlyTensorSpan<T>.Equals(object)' overrides non-obsolete member 'object.Equals(object)'
+#pragma warning disable 8500 // address / sizeof of managed types
+
+namespace System.Numerics.Tensors
+{
+    /// <summary>
+    /// ReadOnlyTensorSpan represents a contiguous region of arbitrary memory. Unlike arrays, it can point to either managed
+    /// or native memory, or to memory allocated on the stack. It is type-safe and memory-safe.
+    /// </summary>
+    [DebuggerTypeProxy(typeof(TensorSpanDebugView<>))]
+    [DebuggerDisplay("{ToString(),raw}")]
+    public readonly ref struct ReadOnlyTensorSpan<T>
+    {
+        /// <summary>A byref or a native ptr.</summary>
+        internal readonly ref T _reference;
+        /// <summary>The number of elements this ReadOnlyTensorSpan contains.</summary>
+        internal readonly nint _flattenedLength;
+        /// <summary>The length of the underlying memory. Can be different than the number of elements in the span.</summary>
+        internal readonly nint _memoryLength;
+        /// <summary>The lengths of each dimension.</summary>
+        internal readonly ReadOnlySpan<nint> _lengths;
+        /// <summary>The strides representing the memory offsets for each dimension.</summary>
+        private readonly ReadOnlySpan<nint> _strides;
+
+        /// <summary>
+        /// Creates a new span over the entirety of the target array.
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <remarks>Returns default when <paramref name="array"/> is null.</remarks>
+        /// <exception cref="ArrayTypeMismatchException">Thrown when <paramref name="array"/> is covariant and array's type is not exactly T[].</exception>
+        public ReadOnlyTensorSpan(T[]? array) : this(array, 0, default, default)
```

Comment:
nit: Just using `default` leaves things a bit ambiguous as to what the type is meant to be, using `ReadOnlySpan<nint>.Empty` (or the empty collection expression, `[]`) is clearer

Doesn't need to be handled in this PR